### PR TITLE
Improve time complexity of IndexOf.

### PR DESCRIPTION
**Pull request by jbialobr**
The complexity was O(n), which is not acceptable for a method
that is used to compare elements in a sorting algorithm.
The new time complexity is O(1).

### CODE REVIEW

File: GitUI/CommitInfo/CommitInfo.cs [hunk 1]
```diff
@@ -711,12 +722,12 @@ protected override void DisposeCustomResources()
 
         private sealed class ItemTpComparer : IComparer<string>
```
Comment (drewnoakes) [inline GitUI/CommitInfo/CommitInfo.cs]: Do you know what this class's name means?

File: GitUI/CommitInfo/CommitInfo.cs [hunk 2]
```diff
@@ -78,7 +78,7 @@ public partial class CommitInfo : GitModuleControl
         private List<string> _branches;
         private string _branchInfo;
         private string _gitDescribeInfo;
-        [CanBeNull] private IList<string> _sortedRefs;
+        [CanBeNull] private IDictionary<string, int> _sortedRefs;
```
Comment (drewnoakes) [inline GitUI/CommitInfo/CommitInfo.cs]: As this is no longer technically "sorted" maybe `_indexByRefName` is a better name for the field.

File: GitUI/CommitInfo/CommitInfo.cs [hunk 3]
```diff
@@ -711,12 +722,12 @@ protected override void DisposeCustomResources()
 
         private sealed class ItemTpComparer : IComparer<string>
         {
-            private readonly IList<string> _otherList;
+            private readonly IDictionary<string, int> _orderDict;
```
Comment (drewnoakes) [inline GitUI/CommitInfo/CommitInfo.cs]: Field/param in this type should match outer type IMO.

File: GitUI/CommitInfo/CommitInfo.cs [hunk 4]
```diff
@@ -726,21 +737,18 @@ public int Compare(string a, string b)
 
                 int IndexOf(string s)
                 {
-                    var head = s.StartsWith("remotes/") ? "refs/" : _prefix;
-                    var tail = s;
-                    var headLength = head.Length;
-                    var length = headLength + s.Length;
-
-                    for (var i = 0; i < _otherList.Count; i++)
+                    if (s.StartsWith("remotes/"))
```
Comment (drewnoakes) [inline GitUI/CommitInfo/CommitInfo.cs]: Can the user of this comparer persist the data in sanitised form so we don't have to allocate temporary strings here?

File: GitUI/CommitInfo/CommitInfo.cs (hunk 1 above)
Comment (jbialobr) [reply GitUI/CommitInfo/CommitInfo.cs]: I have no idea.

File: GitUI/CommitInfo/CommitInfo.cs (hunk 3 above)
Comment (jbialobr) [reply GitUI/CommitInfo/CommitInfo.cs]: I don't know what you mean. Could you explain it?

File: GitUI/CommitInfo/CommitInfo.cs (hunk 4 above)
Comment (jbialobr) [reply GitUI/CommitInfo/CommitInfo.cs]: I don't know. Could you try to convince me that it is worth to care about this? These are short-lived strings and AFAIK the GC deals pretty well with them. The gain in performance caused by avoiding string allocations is almost none if any at all (The old version of this method with string allocations seems to be faster - didn't do a benchmark, just compared freezes times when navigating through blame lines). The gain achieved by replacing List with Dictionary is significant, noticeable with a bare eye. There are many other places where optimization could be done and I would rather optimize them before doing so low level optimizations like avoiding allocations.

File: GitUI/CommitInfo/CommitInfo.cs (hunk 3 above)
Comment (drewnoakes) [reply GitUI/CommitInfo/CommitInfo.cs]: Sure, I mean in this case the field's called `_orderDict` and it's called `_sortedRefs` in the outer scope. They're the same object and it'd be clearer if they had the same name.

Comment (jbialobr) [reply GitUI/CommitInfo/CommitInfo.cs]: Thank you, now it is clear.

File: GitUI/CommitInfo/CommitInfo.cs (hunk 4 above)
Comment (drewnoakes) [reply GitUI/CommitInfo/CommitInfo.cs]: @jbialobr I don't doubt there's already a gain, the implementation looks great. I'm just wary of comparers that allocate. I don't even know how this comparer is used, but if it was impacting perf due to algorithmic complexity, then presumably there were quite a few calls. Right now it's allocating temp strings for each comparison. If the inputs were all canonical (`ref/` form or whatever) then you could eliminate the allocations and the `StartsWith` call.

**Pull request merged**